plugins/tedge_apt_plugin/src/lib.rs

```diff
@@ -103,29 +117,70 @@ fn run_op(apt: AptCli) -> Result<ExitStatus, InternalError> {
+                        include_name.unwrap_or_default(),
+                        include_maintainer.unwrap_or_default()
                     )
                     .as_str(),
                 ) {
-                    Ok(filter) => filter,
+                    Ok(filter) => Some(filter),
                     Err(err) => {
                         eprintln!("tedge-apt-plugin fails to list packages with matching name and maintainer: {err}");
                         std::process::exit(1)
                     }
                 },
             };
 
-            for line in stdout.trim_end().lines() {
-                if filter.is_match(line) {
-                    let (name, version) = get_name_and_version(line);
-                    println!("{name}\t{version}");
+            let exclude_filter = match (&exclude_name, &exclude_maintainer) {
+                (None, None) => None,
+
+                _ => match Regex::new(
+                    format!(
+                        r"(^{}\t.*|^\S+\t\S+\t{}\s+.*)",
+                        exclude_name.unwrap_or_default(),
+                        exclude_maintainer.unwrap_or_default()
+                    )
+                    .as_str(),
+                ) {
+                    Ok(filter) => Some(filter),
+                    Err(err) => {
+                        eprintln!("tedge-apt-plugin fails to list packages with matching name and maintainer: {err}");
+                        std::process::exit(1)
+                    }
+                },
+            };
+
+            let installed_filter = Regex::new(r"install ok installed").unwrap();
+
+            match (include_filter.as_ref(), exclude_filter.as_ref()) {
+                (Some(include_filter), Some(exclude_filter)) => {
+                    for line in stdout.trim_end().lines() {
+                        if installed_filter.is_match(line)
+                            && (include_filter.is_match(line) || !exclude_filter.is_match(line))
+                        {
+                            let (name, version) = get_name_and_version(line);
+                            println!("{name}\t{version}");
+                        }
+                    }
+                }
+                _ => {
+                    let include_filter =
+                        include_filter.unwrap_or_else(|| Regex::new(r".*").unwrap());
+                    let exclude_filter =
+                        exclude_filter.unwrap_or_else(|| Regex::new(r"^$").unwrap());
+                    for line in stdout.trim_end().lines() {
+                        if installed_filter.is_match(line)
+                            && (include_filter.is_match(line) && !exclude_filter.is_match(line))
+                        {
+                            let (name, version) = get_name_and_version(line);
+                            println!("{name}\t{version}");
+                        }
+                    }
                 }
             }
```

Comment:
The `match` is useless as the `_` case simply unwraps the given filters.

```suggestion
                    let include_filter =
                        include_filter.unwrap_or_else(|| Regex::new(r".*").unwrap());
                    let exclude_filter =
                        exclude_filter.unwrap_or_else(|| Regex::new(r"^$").unwrap());
                    for line in stdout.trim_end().lines() {
                        if installed_filter.is_match(line)
                            && (include_filter.is_match(line) && !exclude_filter.is_match(line))
                        {
                            let (name, version) = get_name_and_version(line);
                            println!("{name}\t{version}");
                        }
                    }
```

Reply:
> The `match` is useless as the `_` case simply unwraps the given filters.

Again I was wrong. I missed the use of a disjonction (`||`) vs a conjonction (`&&`).

Reply:
That said, is this really what we expect?

##  Current behavior of this PR

With `include_filter.is_match(line) || !exclude_filter.is_match(line)`, a setting that includes `.*python.*` and excludes `.*lib*.`, would, as expected, emits `python` and `python-httplib` and rejects `libcurl`. But it would also emit `curl` and anything that doesn't match the exclude filter.

## How I understand the priority requirement

To meet the requirement that the `apt.include.*` patterns take precedence over the exclude logic, a better solution would be to ignore all the  `apt.exclude.*` patterns when the user provides an `apt.include.*` pattern. With `include_filter.is_match(line) `, a setting that includes `.*python.*` and excludes `.*lib*.`, would emit `python` as well as `python-httplib`, rejecting `curl` and `libcurl`.

## What I would prefer

I do think the better is to always use the two filters provided by the user with `include_filter.is_match(line) && !exclude_filter.is_match(line)`. A setting that includes `.*python.*` and excludes `.*lib*.`, would emit `python` and reject `python-httplib` as well as `curl` and `libcurl`. To list both  `python` and `python-httplib`, the user has to provide only an include pattern for `.*python.*`. This is what is suggested by https://github.com/thin-edge/thin-edge.io/pull/2869#discussion_r1600560753

Reply:
The main use-case would reflect a real-world scenario:

1. Create a filter to exclude all "lib" values
2. Someone then finds that there is an exception to the "generic exclusion filter", where `lib-curl` needs to be included.

How would this be made possible?

Reply:
I see the following cases:

### Case 1: Explicit exclude pattern

```
NOT (NAME MATCHES EXCLUDE_PATTERN)
```

### Case 2: Include with exclusion exceptions

```
NAME MATCHES INCLUDE_PATTERN
AND NOT (NAME MATCHES EXCLUDE_PATTERN)

```

### Case 3: Exclude with include exceptions

```
NAME MATCHES INCLUDE_PATTERN
OR NOT (NAME MATCHES EXCLUDE_PATTERN
```